### PR TITLE
feat(azure): Add check `defender_auto_provisioning_log_analytics_agent_vms_on`

### PR DESCRIPTION
**Pull request by puchy22**
### Context

Add new check to ensure Auto Provisioning Log Analytics in VMs is On

### Description

Add the new check for Defender service with its respective unit tests

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.metadata.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    "Provider": "azure",
+    "CheckID": "defender_auto_provisioning_log_analytics_agent_vms_on",
+    "CheckTitle": "Ensure that Auto provisioning of 'Log Analytics agent for Azure VMs' is Set to 'On'",
+    "CheckType": [],
+    "ServiceName": "defender",
+    "SubServiceName": "",
+    "ResourceIdTemplate": "",
+    "Severity": "medium",
+    "ResourceType": "AzureDefenderPlan",
+    "Description": "Ensure that Auto provisioning of 'Log Analytics agent for Azure VMs' is Set to 'On'",
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.metadata.json]:
```suggestion
    "Description": "Ensure that Auto provisioning of 'Log Analytics agent for Azure VMs' is Set to 'On'. The Microsoft Monitoring Agent scans for various security-related configurations and events such as system updates, OS vulnerabilities, endpoint protection, and provides alerts.",
```

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.metadata.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    "Risk": "Missing critical security information about your Azure VMs, such as security alerts, security recommendations, and change tracking.",
+    "RelatedUrl": "https://www.trendmicro.com/cloudoneconformity-staging/knowledge-base/azure/SecurityCenter/automatic-provisioning-of-monitoring-agent.html",
+    "Remediation": {
+      "Code": {
+        "CLI": "https://www.trendmicro.com/cloudoneconformity-staging/knowledge-base/azure/SecurityCenter/automatic-provisioning-of-monitoring-agent.html",
+        "NativeIaC": "",
+        "Other": "",
+        "Terraform": "https://docs.bridgecrew.io/docs/ensure-that-azure-defender-is-set-to-on-for-app-service#terraform"
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.metadata.json]: Is this right?

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.metadata.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      },
+      "Recommendation": {
+        "Text": "Ensure comprehensive visibility into possible security vulnerabilities, including missing updates, misconfigured operating system security settings, and active threats, allowing for timely mitigation and improved overall security posture",
+        "Url": ""
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.metadata.json]:
```suggestion
        "Url": "https://learn.microsoft.com/en-us/azure/defender-for-cloud/monitoring-components"
```

File: prowler/providers/azure/services/defender/defender_service.py [hunk 4]
```diff
@@ -37,9 +38,38 @@ def __get_pricings__(self):
...
+                                resource_id=ap.id,
+                                resource_type=ap.type,
+                                auto_provision=ap.auto_provision,
+                            )
+                        }
+                    )
+            except Exception as error:
+                logger.error(f"Subscription name: {subscription}")
+                logger.error(
+                    f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_service.py]:
```suggestion
                logger.error(
                    f"Subscription name: {subscription} -- {error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
```

File: prowler/providers/azure/services/defender/defender_service.py [hunk 5]
```diff
@@ -37,9 +38,38 @@ def __get_pricings__(self):
...
+        return auto_provisioning
 
 
 class Defender_Pricing(BaseModel):
     resource_id: str
     pricing_tier: str
     free_trial_remaining_time: timedelta
+
+class AutoProvisioningSetting(BaseModel):
+    resource_id: str
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_service.py:74]: Maybe it is interesting to store also the setting name. I leave this up to you.

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class defender_auto_provisioning_log_analytics_agent_vms_on(Check):
+    def execute(self) -> Check_Report_Azure:
+        findings = []
+
+        for subscription, auto_provisioning_settings in defender_client.auto_provisioning_settings.items():
+            report = Check_Report_Azure(self.metadata())
+            report.status = "PASS"
+            report.subscription = subscription
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py]: if `subscription` is the `subscription_id` please use that name to be more clear.

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+            report.resource_name = "Defender Auto Provisioning Log Analytics Agents On"
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py]: The `resource_name` maybe should be the `auto_provisioning_settings.name`, what do you think?

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+            report.resource_id = auto_provisioning_settings["default"].resource_id
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py]: Why the `default`? Just curious.

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+            report.status_extended = f"Defenter Auto Provisioning Log Analytics Agents from subscription {subscription} is set to ON."
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py]:
```suggestion
            report.status_extended = f"Defender Auto Provisioning Log Analytics Agents from subscription {subscription} is set to ON."
```

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+            for ap in auto_provisioning_settings:
+                if auto_provisioning_settings[ap].auto_provision != "On":
+                    report.status = "FAIL"
+                    report.status_extended = f"Defenter Auto Provisioning Log Analytics Agents from subscription {subscription} is set to OFF."
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py]:
```suggestion
                    report.status_extended = f"Defender Auto Provisioning Log Analytics Agents from subscription {subscription} is set to OFF."
```

File: tests/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on_test.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
+from unittest import mock
+from uuid import uuid4
+
+from prowler.providers.azure.services.defender.defender_service import AutoProvisioningSetting
+
+AZURE_SUSCRIPTION = str(uuid4())
```
Comment (jfagoagas) [inline tests/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on_test.py]: You can use this value from the `tests/prowler/providers/azure/azure_fixtures.py`. Pedro included that in the previous PR.

**Review by jfagoagas — CHANGES_REQUESTED** [review]
That's really good work @puchy22 👏 

I left some comments just to clarify some parts, but the overall is great!

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py (hunk 8 above)
Comment (puchy22) [reply prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py]: Only checks for default workspace for this check

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+class defender_auto_provisioning_log_analytics_agent_vms_on(Check):
+    def execute(self) -> Check_Report_Azure:
+        findings = []
+
+        for subscription_name, auto_provisioning_settings in defender_client.auto_provisioning_settings.items():
+            report = Check_Report_Azure(self.metadata())
+            report.status = "PASS"
+            report.subscription = subscription_name
+            report.resource_name = auto_provisioning_settings["default"].resource_name
```
Comment (jfagoagas) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.py]: I'm still not sure if we should include here a condition just to review the `default` resource. Let's talk with @sergargar on Monday.

**Review by jfagoagas — DISMISSED** [review]
I'm still not sure if we should include a condition in the check just to only audit the `default` resource because some clients can have different or custom things in Azure and in that case we will generate findings maybe not having to. My vote is to include that filter for now to be 100% sure.

Also this check is going to be deprecated in August 2024 per https://learn.microsoft.com/en-us/azure/azure-monitor/agents/log-analytics-agent

 Let's talk with @sergargar on Monday.

File: prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.metadata.json [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  "SubServiceName": "",
+  "ResourceIdTemplate": "",
+  "Severity": "medium",
+  "ResourceType": "AzureDefenderPlan",
+  "Description": "Ensure that Auto provisioning of 'Log Analytics agent for Azure VMs' is Set to 'On'. The Microsoft Monitoring Agent scans for various security-related configurations and events such as system updates, OS vulnerabilities, endpoint protection, and provides alerts.",
+  "Risk": "Missing critical security information about your Azure VMs, such as security alerts, security recommendations, and change tracking.",
+  "RelatedUrl": "https://www.trendmicro.com/cloudoneconformity-staging/knowledge-base/azure/SecurityCenter/automatic-provisioning-of-monitoring-agent.html",
+  "Remediation": {
+    "Code": {
+      "CLI": "https://www.trendmicro.com/cloudoneconformity-staging/knowledge-base/azure/SecurityCenter/automatic-provisioning-of-monitoring-agent.html",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/defender/defender_auto_provisioning_log_analytics_agent_vms_on/defender_auto_provisioning_log_analytics_agent_vms_on.metadata.json]: Put this link in "Other", please

**Review by MrCloudSec — DISMISSED** [review]
Awesome job!! 🚀

**Review by MrCloudSec — APPROVED** [review]

**Pull request merged**